main/guides/getting-started/contract-rpc.md

```diff
@@ -0,0 +1,410 @@
+[Cosmos SDK endpoint](https://docs.cosmos.network/main/core/grpc_rest) works.
+
+```proto
+message MsgWalletSpendAction {
+    bytes owner = 1;
+    string spend_action = 2;
+}
+```
+
+<!-- TODO: owner field, toAccAddress, base64. working example -->
+
+## Querying VStorage
+
+[VStorage](https://github.com/Agoric/agoric-sdk/tree/master/golang/cosmos/x/vstorage#readme) (for "Virtual Storage") is a key-value store that is
+read-only for clients of the consensus layer.
+From within the JavaScript VM, it is accessed via a `chainStorage` API with a node at each
+key that is write-only; a bit like a `console`.
+
+![vstorage query diagram](./assets/vstorage-brand-q.svg)
+
+The protobuf definition is [agoric.vstorage.Query](https://github.com/Agoric/agoric-sdk/blob/mainnet1B/golang/cosmos/proto/agoric/vstorage/query.proto#L11):
+
+```proto
+service Query {
+  // Return an arbitrary vstorage datum.
+  rpc Data(QueryDataRequest) returns (QueryDataResponse) {
+    option (google.api.http).get = "/agoric/vstorage/data/{path}";
+  }
+
+  // Return the children of a given vstorage path.
+  rpc Children(QueryChildrenRequest)
+    returns (QueryChildrenResponse) {
+      option (google.api.http).get = "/agoric/vstorage/children/{path}";
+  }
+}
+```
+
+We can issue queries using, `agd query ...`:
+
+```sh
+$ agd query vstorage children 'published.agoricNames'
+
+children:
+- brand
+- installation
+- instance
+...
+```
+
+The [Agoric CLI](../agoric-cli/) `follow` command supports vstorage
+query plus some of the marshalling conventions discussed below:
+
+```sh
+$ agoric follow -lF :published.agoricNames.brand
+[
+  [
+    "BLD",
+    slotToVal("board0566","Alleged: BLD brand"),
+  ],
+  [
+    "IST",
+    slotToVal("board0257","Alleged: IST brand"),
+  ],
+...
+]
+```
+
+::: tip vstorage viewer by p2p
+
+The [vstorage-viewer](https://github.com/p2p-org/p2p-agoric-vstorage-viewer) contributed by p2p is often _very_ handy:
+
+[![vstorage viewer screenshot](https://user-images.githubusercontent.com/150986/259798595-40cd22f0-fa01-43a9-b92a-4f0f4813a4f6.png)](https://p2p-org.github.io/p2p-agoric-vstorage-viewer/#https://devnet.rpc.agoric.net/|published,published.agoricNames|)
+
+:::
+
+## Specifying Offers
+
+Recall that for an agent within the JavaScript VM,
+[E(zoe).offer(...)](../../reference/zoe-api/zoe.md#e-zoe-offer-invitation-proposal-paymentkeywordrecord-offerargs) takes an `Invitation` and optionally a `Proposal` with `{ give, want }`, a `PaymentKeywordRecord`, and `offerArgs`; it returns a `UserSeat` from which we can [getPayouts()](../../reference/zoe-api/user-seat.md#e-userseat-getpayouts).
+
+![Zoe API diagram, simplified](./assets/zoe-simp.svg)
+
+In the Smart Wallet architecture, a client uses an `OfferSpec` to
+tell its `SmartWallet` how to conduct an offer.
+It includes an `invitationSpec` to say what invitation to pass to Zoe. For example:
+
+<<< @/snippets/test-marshal.js#exInvitationSpec
```

Comment:
forward reference to how we got `instance`?

explain that the code is excerpted from some snippets that provide some stuff in ways not shown?